File: src/DataScience/Octave/proxy.m
function proxy()
  backend_created = 0;
  fetches = 0;
  cache = containers.Map('KeyType', 'char', 'ValueType', 'char');
  subject_created = false;

  first = get_document(42);
  second = get_document(42);
  printf('backend=%d;fetches=%d;first=%s;second=%s\n', backend_created, fetches, first, second);

  function value = get_document(id)
    key = num2str(id);
    if isKey(cache, key)
      value = cache(key);
      return;
    endif

    if !subject_created
      backend_created = backend_created + 1;
      subject_created = true;
    endif

    fetches = fetches + 1;
    value = sprintf('doc(%s)', key);
    cache(key) = value;
  endfunction
endfunction

proxy();

File: src/Web/Dart/proxy.dart
abstract interface class DocumentStore {
  String fetch(int id);
}

final class DocumentBackend implements DocumentStore {
  int fetches = 0;

  @override
  String fetch(int id) {
    fetches += 1;
    return 'doc($id)';
  }
}

final class DocumentProxy implements DocumentStore {
  DocumentBackend? _backend;
  final Map<int, String> _cache = <int, String>{};

  int get backendCount => _backend == null ? 0 : 1;
  int get fetches => _backend?.fetches ?? 0;

  @override
  String fetch(int id) {
    return _cache.putIfAbsent(id, () {
      _backend ??= DocumentBackend();
      return _backend!.fetch(id);
    });
  }
}

void main() {
  final proxy = DocumentProxy();
  final first = proxy.fetch(42);
  final second = proxy.fetch(42);
  print(
    'backend=${proxy.backendCount};fetches=${proxy.fetches};first=$first;second=$second',
  );
}
